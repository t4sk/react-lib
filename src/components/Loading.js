import React from "react"
import PropTypes from "prop-types"
import { Message, Button } from "semantic-ui-react"

export function Loading(props) {
  if (props.loading) {
    return props.renderLoading()
  }

  if (props.error) {
    return (
      <Message negative>
        <p>{props.error}</p>
        <Button onClick={props.onClickRetry}>Retry</Button>
      </Message>
    )
  }

  return props.children
}

Loading.propTypes = {
  loading: PropTypes.bool.isRequired,
  error: PropTypes.string.isRequired,
  children: PropTypes.node,
  onClickRetry: PropTypes.func.isRequired,
  renderLoading: PropTypes.func.isRequired,
}

Loading.defaultProps = {
  renderLoading: () => <div>Loading...</div>,
}

export default Loading
