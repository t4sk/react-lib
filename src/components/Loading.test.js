import React from "react"
import { shallow } from "enzyme"
import { Loading } from "./Loading"

const children = <div>children</div>
const props = {
  loading: false,
  error: "",
  onClickRetry: jest.fn(),
}

test("it renders", () => {
  const component = shallow(<Loading {...props}>{children}</Loading>)

  expect(component).toMatchSnapshot()
})

test("loading", () => {
  const component = shallow(
    <Loading {...props} loading={true}>
      {children}
    </Loading>
  )

  expect(component).toMatchSnapshot()
})

test("error", () => {
  const component = shallow(
    <Loading {...props} error="Error">
      {children}
    </Loading>
  )

  expect(component).toMatchSnapshot()
})

test("renderLoading", () => {
  const renderLoading = () => <div>Custom loading message...</div>
  const component = shallow(
    <Loading {...props} error="Error" renderLoading={renderLoading}>
      {children}
    </Loading>
  )

  expect(component).toMatchSnapshot()
})

test("onClickRetry", () => {
  const component = shallow(
    <Loading {...props} error="Error">
      {children}
    </Loading>
  )

  component.find("Button").simulate("click")

  expect(props.onClickRetry.mock.calls.length).toEqual(1)
})
