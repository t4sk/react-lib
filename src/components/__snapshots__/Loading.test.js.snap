// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error 1`] = `
<Message
  negative={true}
>
  <MessageHeader>
    Error
  </MessageHeader>
  <p>
    Error
  </p>
  <Button
    as="button"
    onClick={[MockFunction]}
  >
    Retry
  </Button>
</Message>
`;

exports[`it renders 1`] = `
<div>
  children
</div>
`;

exports[`loading 1`] = `
<div>
  Loading...
</div>
`;

exports[`renderLoading 1`] = `
<Message
  negative={true}
>
  <MessageHeader>
    Error
  </MessageHeader>
  <p>
    Error
  </p>
  <Button
    as="button"
    onClick={[MockFunction]}
  >
    Retry
  </Button>
</Message>
`;
